### DISCUSSION TITLE
IKEv2 No Reconnect after modem disconnect

### DISCUSSION THREAD

**Discussion by Gerbaum** [strongswan/strongswan#401] [Q&A]
Hi,
I got an Ubuntu 20.04 Kernel 5.4 Armv6 device with 4G Modem as a roadwarrior client and have established a working ikev2 connection to my Server managed by a mikrotik router.

Connection works without problems, rekeying every 3 hours. 

Linux strongSwan Version is U5.8.2/K5.4.45-g0d81aa75557d

Problem occurs when 4G modem losses connection and connects again.

I have set auto=add, followed by a "ipsec up test" skript with a 30s delay after boot, because auto=start tried too soon to initiate ipsec without 4G connection, which led to a non working connection.

**ipsec.conf**

 `conn test`
 `keyexchange=ikev2`
 `ike=aes128-sha1-modp2048`
 `esp=aes128-sha1`
 `left=%defaultroute`
 `leftsourceip=%modeconfig`
 `leftcert=cert_export_Cert.crt`
 `leftid=CertTest`
 `leftfirewall=yes`
 `right=x.x.x.x`
 `rightsubnet=0.0.0.0/0`
 `rightid="x.x.x.x"`
 `dpddelay=30s`
 `dpdaction=clear`
 `auto=add`

My modems works as a router, so it has internally a public IP, and gives out a privat IP to my device(192.168.1.100)

**Working Condition:**

```
Listening IP addresses:
  192.168.1.100
Connections:
        test:  %any...x.x.x.x  IKEv2, dpddelay=30s
        test:   local:  [CN=CertTest] uses public key authentication
        test:    cert:  "CN=CertTest"
        test:   remote: [x.x.x.x] uses public key authentication
        test:   child:  dynamic === 0.0.0.0/0 TUNNEL, dpdaction=clear
Security Associations (1 up, 0 connecting):
        test[7]: ESTABLISHED 100 minutes ago, 192.168.1.100[CN=CertTest]...x.x.x.x[x.x.x.x]
        test[7]: IKEv2 SPIs: 0194e1bf0c7a268b_i* 5c32a5c185628f75_r, public key reauthentication in 63 minutes
        test[7]: IKE proposal: AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
        test{27}:  INSTALLED, TUNNEL, reqid 7, ESP in UDP SPIs: c641797f_i 0f4dd105_o
        test{27}:  AES_CBC_128/HMAC_SHA1_96, 10220 bytes_i (125 pkts, 0s ago), 17104 bytes_o (96 pkts, 8s ago), rekeying in 38 minutes
        test{27}:   192.168.11.254/32 === 0.0.0.0/0
```

**1 min after disconnect, non working**

```
Listening IP addresses:
  192.168.1.100
Connections:
        test:  %any...x.x.x.x  IKEv2, dpddelay=30s
        test:   local:  [CN=CertTest] uses public key authentication
        test:    cert:  "CN=CertTest"
        test:   remote: [x.x.x.x] uses public key authentication
        test:   child:  dynamic === 0.0.0.0/0 TUNNEL, dpdaction=clear
Security Associations (1 up, 0 connecting):
        test[7]: ESTABLISHED 2 hours ago, 192.168.1.100[CN=CertTest]...x.x.x.x[x.x.x.x]
        test[7]: IKEv2 SPIs: 0194e1bf0c7a268b_i* 5c32a5c185628f75_r, public key reauthentication in 11 minutes
        test[7]: IKE proposal: AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
        test{28}:  INSTALLED, TUNNEL, reqid 7, ESP in UDP SPIs: cda3fb0e_i 07dae64e_o
        test{28}:  AES_CBC_128/HMAC_SHA1_96, 18252 bytes_i (124 pkts, 116s ago), 5952 bytes_o (81 pkts, 116s ago), rekeying in 31 minutes
        test{28}:   192.168.11.254/32 === 0.0.0.0/0
```

My router shows both installed Sas, and I can see traffic passing through the tunnel (at least incoming traffic) some minutes later:

```
Security Associations (1 up, 0 connecting):
        test[7]: ESTABLISHED 2 hours ago, 192.168.1.100[CN=CertTest]...x.x.x.x[x.x.x.x]
        test[7]: IKEv2 SPIs: 0194e1bf0c7a268b_i* 5c32a5c185628f75_r, public key reauthentication in 3 minutes
        test[7]: IKE proposal: AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
        test{28}:  INSTALLED, TUNNEL, reqid 7, ESP in UDP SPIs: cda3fb0e_i 07dae64e_o
        test{28}:  AES_CBC_128/HMAC_SHA1_96, 26820 bytes_i (277 pkts, 623s ago), 5952 bytes_o (81 pkts, 623s ago), rekeying in 22 minutes
        test{28}:   192.168.11.254/32 === 0.0.0.0/0
```

But the tunnel will only come up if I manually `ipsec down test` and then` ipsec up test`, or I'll have to wait 3 hours till next rekeying.

**Working condition after manual `ipsec down test` followed by` ipsec up test`**

```
Listening IP addresses:
  192.168.1.100
Connections:
        test:  %any...x.x.x.x  IKEv2, dpddelay=30s
        test:   local:  [CN=CertTest] uses public key authentication
        test:    cert:  "CN=CertTest"
        test:   remote: [x.x.x.x] uses public key authentication
        test:   child:  dynamic === 0.0.0.0/0 TUNNEL, dpdaction=clear
Security Associations (1 up, 0 connecting):
        test[9]: ESTABLISHED 14 seconds ago, 192.168.1.100[CN=VentanaTest]...x.x.x.x[x.x.x.x]
        test[9]: IKEv2 SPIs: 005a60483ed67e81_i* 23169b5276718829_r, public key reauthentication in 2 hours
        test[9]: IKE proposal: AES_CBC_128/HMAC_SHA1_96/PRF_HMAC_SHA1/MODP_2048
        test{30}:  INSTALLED, TUNNEL, reqid 9, ESP in UDP SPIs: c8beb6fc_i 01fa1322_o
        test{30}:  AES_CBC_128/HMAC_SHA1_96, 112 bytes_i (2 pkts, 3s ago), 112 bytes_o (2 pkts, 3s ago), rekeying in 45 minutes
        test{30}:   192.168.11.254/32 === 0.0.0.0/0
```

Konfiguration "auto=route" creates duplicate SAs for me after first init, but tunnel is working too.
But after modem disconnect, my tunnel wont work either, just incoming bytes.
Ipsec down und ipsec up helps here too

Why doesn't it reconnect itself?

**Comment by tobiasbrunner** [comment]
> **1 min after disconnect, non working**

In what way? You even write traffic is passed through the tunnel. So what's not working? What about traffic counters (the output you posted above doesn't make much sense as the working condition output seems to be taken _after_ the "broken" one)? And what about DPDs/MOBIKE (check the logs)?

**Comment by Gerbaum** [comment]
Hi,
after my disconnect my modem auto connects again and even preserves its public ip address.
Internet connection works, but not through tunnel any longer.

The tunnel is still active and I see packets incoming from my server, but no packets outgoing. 
I guess somehow my outgoing routing policies doesn't work any longer. 

Is there any command to check this?

**Reply by tobiasbrunner** [reply]
You can check the route in table 220 (`ip route list table 220`), installing the virtual IP on a different interface (e.g. `lo`) might also help (`charon.install_virtual_ip_on`) in case the interface disappears completely.

**Comment by Gerbaum** [comment]
You are right, after disconnect there is no input in table 220.  I don't know why it gets deleted, cause `mmcli -m 0 --simple-disconnect`  just disconnects my carrier connection inside my modem, interface usb0 itself won't be touched.

Is there a workaround  to recognize an empty table 220 and refresh ist?

**Reply by tobiasbrunner** [reply]
Try the option I mentioned (virtual IPs currently won't get reinstalled if they disappear, so make sure they are on a stable interface). Routes are reinstalled, but if the virtual IP is gone that might fail. You can instease the log level for _knl_ to 2 to see what's going on with the interface.
